pkg/datagatherer/istio/istio.go

```diff
@@ -0,0 +1,156 @@
+// Package isito provides a datagatherer for Istio.
+package isito
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"strings"
+	"time"
+
+	"github.com/jetstack/preflight/pkg/datagatherer/k8s"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers"
+	"istio.io/istio/galley/pkg/config/analysis/local"
+	"istio.io/istio/pkg/config/resource"
+	istioSchema "istio.io/istio/pkg/config/schema"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/jetstack/preflight/pkg/datagatherer"
+	dgerror "github.com/jetstack/preflight/pkg/datagatherer/error"
+)
+
+// Config is the configuration for the Istio DataGatherer.
+type Config struct {
+	// KubeConfigPath is the path to the kubeconfig file. If empty, will assume it runs in-cluster.
+	KubeConfigPath string `yaml:"kubeconfig"`
+	// IstioNamespace is the Namespace for Istio components, this is usually `istio-system`.
+	IstioNamespace string `yaml:"istio-namespace"`
+	// ExcludeNamespaces is a list of namespaces to exclude.
+	ExcludeNamespaces []string `yaml:"exclude-namespaces"`
+	// IncludeNamespaces is a list of namespaces to include.
+	IncludeNamespaces []string `yaml:"include-namespaces"`
+	// Resources is a list of GroupVersionResources to collect for Istio analysis.
+	Resources []schema.GroupVersionResource
+}
+
+// validate validates the configuration.
+func (c *Config) validate() error {
+	var errors []string
+
+	if len(c.ExcludeNamespaces) > 0 && len(c.IncludeNamespaces) > 0 {
+		errors = append(errors, "cannot set excluded and included namespaces")
+	}
+
+	if len(errors) > 0 {
+		return fmt.Errorf(strings.Join(errors, ", "))
+	}
+
+	return nil
+}
+
+// DataGatherer is a DataGatherer for Istio.
+type DataGatherer struct {
+	sa       *local.SourceAnalyzer
+	dynamics []datagatherer.DataGatherer
+}
+
+// NewDataGatherer creates a new DataGatherer for a cluster.
+func (c *Config) NewDataGatherer(ctx context.Context) (datagatherer.DataGatherer, error) {
+	if err := c.validate(); err != nil {
+		return nil, err
+	}
+
+	// If the Istio namespace is not set then default it to 'istio-system'.
+	istioNamespace := c.IstioNamespace
+	if istioNamespace == "" {
+		istioNamespace = "istio-system"
+	}
+
+	// If a list of resources is not set then use the data gatherer default list.
+	gvrs := c.Resources
+	if len(gvrs) == 0 {
+		gvrs = defaultGVRs
+	}
+
+	// Create a list of Kubernetes dynamic data gatherers to fetch all the required resources for Istio analysis.
+	var dynamics []datagatherer.DataGatherer
```

Comment:
Just a suggestion, i'd rather give this a more descriptive name.

```suggestion
	var dynamicDataGatherers []datagatherer.DataGatherer
```